# Manual Review of Heads and Features

The manual annotations are intended to serve 2 roles: 1. to evaluate the accuracy of the head assignments on time phrases in the BHSA data, 2. to evaluate the time phrase structure in the BHSA dataset, and 3. to gain hands-on exposure to the kinds of time phrases in the dataset. This process consisted of comparing the selected heads against the surface text of the time phrase, and of reading the time phrases in the context of a verse when questions or anomalies arose. The annotation process consisted of marking a given time phrase as "g" for "good," "b" for "bad," and "?" for questionable cases. These classifications refer to both head assignments and internal structuring of the time phrases in the BHSA. The markings are often accompanied with notes: for bad or questionable entries the note explains what is wrong, for good entries the note might describe an interesting phenomenon, in some cases it might give a "light caution" about a given phrase. 

The outcome of this annotation process suggests that a custom database is necessary to consistently represent time phrases: there are many cases in BHSA time phrases where the phrase is cleft into 2 adjacent parts whereas in the majority of the data they are kept together. This is an inconsistency that should be solved. In other cases, the notion of "phrase" is not broad enough to encompass the full range of expressions that can mark time. For instance, several time phrases are split off from the infinitives they direct, in which the infinitive is an event. This is because the ETCBC method defines the infinitive event as a clause, operating at a different hierarchical level than the phrase; phrases are defined as strictly non-predicative. In the framework of Construction Grammar (Goldberg, *Constructions*, 1995, Croft, *Radical Construction Grammar*, 2001) preferred by this study, the link between the two forms is likewise considered a constructional object. Indeed, many phrases in this dataset refer to event-like nouns, which are from a syntactical perspective non-predicative; במות "in the death of" is a very frequent example, but other cases include ביום ישׁועה "in the day of salvation," which assumes a salvation event. If the difference between time phrases and time "clauses" is seen as merely an incremental difference rather than categorical, then one can apply a unified strategy in analyzing these cases. 

### Load TF / BHSA Data

In [22]:
import collections, random, csv
import pandas as pd
from IPython.display import display
from tf.fabric import Fabric
from tf.app import use
pd.set_option('display.max_colwidth', 0)  # configure DataFrame to show full notes with no truncation

locs = {'bhsa':'~/text-fabric-data/etcbc/bhsa/tf/c',
        'heads':'~/github/etcbc/heads/tf/c',
        'custom':'~/github/csl/time_collocations/data/tf'}

# load BHSA
TF = Fabric(locations=locs.values())
api = TF.load('''

vs vt pdp gloss lex typ number  prs
g_cons_utf8 nu mother st language ls rela
obj_prep sem_set head nhead
note function

''')

A = use('bhsa', api=api, hoist=globals(), silent=True)

This is Text-Fabric 7.7.3
Api reference : https://annotation.github.io/text-fabric/Api/Fabric/

121 features found and 4 ignored
  0.00s loading features ...
   |     0.00s Not enough info for structure in otext, structure functionality will not work
   |     0.18s B g_cons_utf8          from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.12s B lex                  from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.10s B vs                   from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.10s B vt                   from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.10s B pdp                  from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.12s B gloss                from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.17s B typ                  from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.18s B number               from /Users/cody/text-fabric-data/etcbc/bhsa/tf/c
   |     0.09s B prs                  from /U

In [3]:
def tokenPhrase(phrasenode):
    '''Tokenizes a phrase with
    dot-separated words.
    input: phrase node number
    output: token string'''
    words = [(F.g_cons_utf8.v(w) if F.lex.v(w) != 'H' else 'ה') for w in L.d(phrasenode, 'word')]
    return '.'.join(words)

def tokenHeads(headslist):
    '''same as tokenPhrase but with list of head word nodes'''
    return '.'.join((F.g_cons_utf8.v(w) if F.lex.v(w) != 'H' else 'ה') for w in headslist)

## Phrase Tokens and Phrase Heads
This search counts all of the discrete time phrase tokens in Hebrew and gathers data about their heads. This data is exported to a spreadsheet for manual inspection. Per every token, a key of its heads is saved into a dictionary, linked to a list of phrase nodes. Tokens that have more than 1 head are suspicious, since the surface form is the same. All other tokens will be exported with their standard heads for inspection.

In [4]:
tp_heads = collections.defaultdict(lambda: collections.defaultdict(list))
tp_nheads = collections.defaultdict(lambda: collections.defaultdict(list))
tp_count = collections.Counter()

tps = A.search('''

phrase function=Time
/with/
    word language=Hebrew
/-/

''', shallow=True)

for tp in tps:
    token = tokenPhrase(tp) # tokenize surface of phrase
    heads_token = tokenHeads(E.head.t(tp)) # get heads
    nheads_token = tokenHeads(E.nhead.t(tp)) # get semantic heads
    tp_heads[token][heads_token].append(tp) # look at its heads
        
    # only populate nheads with PP phrases, since nhead feature for NP is exactly the same
    if F.typ.v(tp) == 'PP':
        tp_nheads[token][nheads_token].append(tp)
        
    tp_count[token] += 1
    
suspect_heads = [tp for tp in tp_heads if len(tp_heads[tp]) > 1]
suspect_nheads = [tp for tp in tp_nheads if len(tp_nheads[tp]) > 1]

print(f'total phrase tokens 2 head mappings: {len(tp_heads)}')
print(f'total phrase tokens 2 nhead mappings: {len(tp_nheads)}')
print(f'total suspect heads: {len(suspect_heads)}')
print(f'total suspect nheads {len(suspect_nheads)}')

  1.15s 3953 results
total phrase tokens 2 head mappings: 1164
total phrase tokens 2 nhead mappings: 888
total suspect heads: 0
total suspect nheads 0


### Compile Manual Inspection Spreadsheet

Two spreadsheets are exported: a `tp_heads.csv` and a `tp_nheads.csv`. Two versions are needed since `nhead` ("nominal" head, which is probably better-named as "semantic head") is different in cases of prepositional phrases. The `nhead` and `head` features on phrases thus have a lot of overlap, but are not identical. The `tp_nheads.csv` contains only prepositional phrases, since their `nhead` (an object of the prep.) is always different from the `head`.

After export, the files are opened in Numbers and manually annotated with a series of tags. It is then re-saved as the same .csv file. The data is exported below, and the results of the manual annotations are described in the subsequent section.

The column `token` contains a tokenized string of the entire phrase. The column `(n)heads_token` contains a dot-separated list of heads assigned to the phrase (e.g. ב.ב are two heads, both ב, for a single phrase). `ex` and `ex_ref` contain an example and example verse reference.

In [5]:
tp_heads_data = []
tp_nheads_data = []
data_header = ['token', '(n)heads_token', 'freq', 'mark', 'note', 'ex_ref', 'ex', 'ex_node', 'ex_verse']

for htp, nhtp in zip(tp_heads.keys(), tp_nheads.keys()):
    head = next(tp for tp in tp_heads[htp])
    nhead = next(tp for tp in tp_nheads[nhtp])
    head_ex = random.choice(tp_heads[htp][head])
    nhead_ex = random.choice(tp_nheads[nhtp][nhead])
    
    head_ref, nhead_ref = ['{} {}:{}'.format(*T.sectionFromNode(ex)) for ex in (head_ex, nhead_ex)]
    head_txt, nhead_txt = [T.text(ex) for ex in (head_ex, nhead_ex)]
    head_verse, nhead_verse = [T.text(L.u(ex, 'verse')[0]) for ex in (head_ex, nhead_ex)]
    
    heads_data = [htp, head, tp_count[htp], '', '', head_ref, head_txt, head_ex, head_verse]
    nheads_data = [nhtp, nhead, tp_count[nhtp], '', '', nhead_ref, nhead_txt, nhead_ex, nhead_verse]
    tp_heads_data.append(heads_data)
    tp_nheads_data.append(nheads_data)
    
tp_heads_data, tp_nheads_data = sorted(tp_heads_data), sorted(tp_nheads_data)

In [6]:
# !! Commented out to prevent writing over the manually added data !!

# with open('manual_curation/tp_heads.csv', 'w') as outfile:
#     writer = csv.writer(outfile)
#     writer.writerow(data_header)
#     writer.writerows(tp_heads_data)
    
# with open('manual_curation/tp_nheads.csv', 'w') as outfile:
#     writer = csv.writer(outfile)
#     writer.writerow(data_header)
#     writer.writerows(tp_nheads_data)

## Results of the Manual Inspection

In [10]:
head_anno = pd.read_csv('tp_heads.csv').fillna('')
nhead_anno = pd.read_csv('tp_nheads.csv').fillna('')

Cases are marked with one of three tags, `g` (good), `?` (questionable), or `b` (bad), based on whether the selected head/nhead is correct. The table below shows the breakdown of the manual annotations.

In [25]:
head_marks = pd.DataFrame.from_dict(collections.Counter(head_anno.mark), orient='index')
head_marks.index = ['good', 'questionable', 'bad']
head_marks['%'] = head_marks[0] / head_marks[0].sum()
head_marks = head_marks.sort_values(by=0, ascending=False)

nhead_marks = pd.DataFrame.from_dict(collections.Counter(nhead_anno.mark), orient='index')
nhead_marks.index = ['good', 'questionable', 'bad']
nhead_marks['%'] = nhead_marks[0] / nhead_marks[0].sum()
nhead_marks = nhead_marks.sort_values(by=0, ascending=False)

print('Head Marks')
display(head_marks)
print('')
print('N-Head Marks')
display(nhead_marks)

Head Marks


0         %
good          1136  0.970111
questionable  21    0.017933
bad           14    0.011956


N-Head Marks


0         %
good          802  0.897092
questionable  53   0.059284
bad           39   0.043624

97% of cases in head marks are rated good. N-heads shows a lower rate at 90%.

### Fixing Bad Cases

The bad cases are all exported to `head_fixes.csv` and `nhead_fixes.csv`. I add a new column, "fix", which contains a simple note proposing what solution should be adopted to remedy the bad case. Those solutions are broken down and explained below.

Export the spreadsheet for annotating proposed fixes to these problems.

**NB that the export is now commented out to prevent overwrite**

In [31]:
# head_anno[head_anno['mark'] == 'b'].to_csv('head_fixes.csv')

**The fixes are added and re-imported below.**

The solutions under 'fix' break down into the following:

* `function2` - these are cases which need to be fixed by reassigning the function of the phrase from `Time` to something else
* `define construction` - these are cases where a new object is needed to handle complexity above or below the phrase level
* `is good` - upon further inspection these cases are not erroneous

For particular cases there is additional information contained in the "note" column about the specific problem.

In [40]:
headfixes = pd.read_csv('head_fixes.csv').fillna('')
headfixes

Unnamed: 0  \
0   214          
1   251          
2   710          
3   858          
4   859          
5   865          
6   949          
7   960          
8   961          
9   962          
10  965          
11  970          
12  976          
13  1097         

                                                              token  \
0   ב.ה.שׁמים.ו.ב.ה.ארץ                                               
1   ב.הר.ציון.ו.ב.ירושׁלם                                             
2   כל.ימי.כורשׁ.מלך.פרס.ו.עד.מלכות.דריושׁ.מלך.פרס                    
3   מ.ה.יום.ה.הוא.ו.הלאה                                              
4   מ.ה.יום.ה.הוא.ו.מעלה                                              
5   מ.יום.ה.שׁמיני.ו.הלאה                                             
6   מהלך.יום.אחד                                                      
7   מן.ה.יום.ה.זה                                                     
8   מן.ה.יום.ה.זה.ו.מעלה                                              
9   מן.ה.יום.ה.זה.ו.מעלה.מ.יום.עשׂרים.ו.ארבעה.ל.ה.תשׁיעי.ל.מן.ה.יום   
10  מן.ה.יום.ו.הלאה.ל.דרתיכם                                          
11  מן.ה.עולם.עד.ה.עולם                                               
12  מספר.ה.ימים.שׁלשׁ.מאות.ו.תשׁעים.יום                               
13  שׁבת.שׁבת                                                         

   (n)heads_token  freq mark  \
0   ב.ב            1     b     
1   ב.ב            1     b     
2   ימי.מלכות      1     b     
3   מ              1     b     
4   מ              2     b     
5   מ              1     b     
6   מהלך           1     b     
7   מן             1     b     
8   מן             1     b     
9   מן.ל           1     b     
10  מן             1     b     
11  מן             1     b     
12  ימים           1     b     
13  שׁבת           1     b     

                                                                                                                                                                                                                                                                                                                                                         note  \
0   bad phrase function (not a time phrase); good heads                                                                                                                                                                                                                                                                                                         
1   not a time phrase; heads are good though                                                                                                                                                                                                                                                                                                                    
2   This TP should contain an NP specified by a PP with waw connection. The heads algorithm appears to have selected past the prep in this case. That is the expected behavior, but it has a bad effect here. There may be other cases like this, which could be isolated by finding independent preps within NP phrases and checking their head assignments.   
3   This TP misses והלאה; my suspicion is that these words are being marked as spec or atr when they should be para.                                                                                                                                                                                                                                            
4   This is like הלאה but with מעלה; m. should be picked up as a coordinated noun. To prove this, it would be helpful to show comparable examples with waw connections.                                                                                                                                                                                         
5   another case of הלאה not encoded                                

Now the nhead problems...

Export a spreadsheet for annotating proposed fixes to these problems.

**NB that the export is now commented out to prevent overwrite**

In [32]:
# nhead_anno[nhead_anno['mark'] == 'b'].to_csv('nhead_fixes.csv')

**The fixes are added and re-imported below.**

There is a new "fix" tag:

* head fixed - this issue has now been solved by re-running the [heads pipeline](https://github.com/ETCBC/heads) with corrections/adjustments

All other tags are identical in meaning to those found in `headfixes`.

In [39]:
nheadfixes = pd.read_csv('nhead_fixes.csv').fillna('')
nheadfixes

Unnamed: 0  \
0   10           
1   38           
2   39           
3   40           
4   41           
5   42           
6   59           
7   113          
8   119          
9   134          
10  179          
11  189          
12  223          
13  226          
14  242          
15  374          
16  401          
17  437          
18  442          
19  445          
20  446          
21  447          
22  448          
23  481          
24  487          
25  488          
26  489          
27  508          
28  510          
29  513          
30  514          
31  522          
32  531          
33  536          
34  559          
35  560          
36  563          
37  567          
38  569          
39  571          
40  579          
41  624          
42  667          
43  686          
44  713          
45  714          
46  720          
47  727          
48  787          
49  793          
50  807          
51  809          
52  810          

                                                              token  \
0   אחרי                                                              
1   אחריה                                                             
2   אחריהם                                                            
3   אחריהן                                                            
4   אחריו                                                             
5   אחריך                                                             
6   ב.אחרית.מלכותם                                                    
7   ב.ה.חמשׁי                                                         
8   ב.ה.יום.ה.הוא.ב.יום                                               
9   ב.ה.יום.ה.שׁמיני.ו.הלאה                                           
10  ב.ה.רביעי                                                         
11  ב.ה.שׁמים.ו.ב.ה.ארץ                                               
12  ב.ה.שׁשׁי                                                         
13  ב.הר.ציון.ו.ב.ירושׁלם                                             
14  ב.חמשׁה.עשׂר.ל.ה.חדשׁ                                             
15  ב.כל                                                              
16  ב.עוד.לילה                                                        
17  ב.רב.ה.ימים                                                       
18  ב.שׁבעה.ל.ה.חדשׁ                                                  
19  ב.שׁלושׁה.ו.עשׂרים.בו                                             
20  ב.שׁלושׁה.עשׂר.יום.בו                                             
21  ב.שׁלושׁה.עשׂר.יום.בו.ב.ה.יום                                     
22  ב.שׁלושׁה.עשׂר.ל.חדשׁ.שׁנים.עשׂר                                  
23  ב.שׁנת.חמשׁים.ו.שׁתים.שׁנה.ל.עזריה.מלך.יהודה                      
24  ב.שׁנת.עשׂרים.ו.שׁבע.שׁנה.ל.ירבעם.מלך.ישׂראל                      
25  ב.שׁנת.עשׂרים.ו.שׁלשׁ.שׁנה.ל.ה.מלך.יהואשׁ                         
26  ב.שׁנת.עשׂרים.ו.שׁלשׁ.שׁנה.ל.יואשׁ.בן.אחזיהו.מלך.יהודה            
27  ב.שׁנת.שׁלשׁים.ו.שׁבע.שׁנה.ל.יואשׁ.מלך.יהודה                      
28  ב.שׁנת.שׁלשׁים.ו.שׁמנה.שׁנה.ל.עזריהו.מלך.יהודה                    
29  ב.שׁנת.שׁלשׁים.ו.תשׁע.שׁנה.ל.עזיה.מלך.יהודה                       
30  ב.שׁנת.שׁלשׁים.ו.תשׁע.שׁנה.ל.עזריה.מלך.יהודה                      
31  ב.שׁנת.שׁשׁ.מאות.שׁנה.ל.חיי.נח                                    
32  ב.שׁנת.שׁתים.עשׂרה.שׁנה.ל.יורם.בן.אחאב.מלך.ישׂראל                 
33  ב.תחלת                                                            
34  כ.ב.ה.ראשׁנה                                                      
35  כ.ב.ה.תחלה                                                        
36  כ.ה.עת                                                            
37  כ.חצת.ה.לילה                                                      
38  כ.יום.מדין                                                        
39  כ.ימות.שׁנות                                                      
40  כ.עת                                                              
41  ל.ימים.מ.ימים                                

## Discussion on Fixes

There are a lot of cases marked as `define construction`. These can be broken down into N main groups:

* missing מעלה/הלאה and equivalents.
* quantified שׁנת "year of" phrases wherein the construct year is followed by a number and another instance of שׁנה. These are not properly missing heads cases, and could potentially be renamed as "cautions" rather than "bad." These particular cases will be handled in [chunking](../pipeline/chunking.py) which is run in the [pipeline](../pipeline/pipeline.ipynb). They are managed by treating the entire quantified phrase as a quant chunk object. That then allows the analysis to examine the quant chunk in its entirety in relation to שׁנת.
* מן...עד and מן...ועד. These are related to the first cases identified above. These cases are interesting because those with the waw should probably be marked as proper heads. But they are very similar to those without waw connections. The situation is perhaps similar to other cases when the BHSA marks an item as "parallel" (i.e. an independent element co-occurring with another independent element) but is not coordinated by waw. **These may represent cases that ought to be corrected in the heads pipeline**. At the same time, they are semantically determined. In other words, this problem might indeed be specific to this pattern. I need to do further work to find out the most consistent way to handle these cases. In the next stage, I will be analyzing complex time adverbial constructions, i.e. those such as these cases which have more than one preposition or time word. That stage will require that these cases be fully accounted for.
* the case in Ezra 4:5 (phrase 881852) where מלכות is selected instead of עד may represent a problem in the heads pipeline as well. This is a case where the phrase as a whole is a NP that contains a coordinate PP. The PP is marked as a separate phrase atom contained in the whole NP. **This case may be systemic across the heads dataset. I need to investigate what can be done to solve this in the heads pipeline.** 

### Looking at Questionable Cases

In [11]:
head_anno[head_anno['mark'] == '?']

token (n)heads_token  freq mark  \
13    אחרי                                    אחרי           4     ?     
159   ב.ה.יום.ה.שׁמיני.ו.הלאה                 ב              1     ?     
245   ב.ה.שׁנים.עד.שׁנת.ה.יבל                 ב              1     ?     
252   ב.זמניהם                                ב              1     ?     
651   ימים.רבים.שׁמונים.ו.מאת.יום             ימים           1     ?     
802   ל.עת.יום.ב.יום                          ל.ב            1     ?     
830   ל.רוח.ה.יום                             ל              1     ?     
864   מ.יום.אל.יום                            מ              2     ?     
876   מ.ימי.מלכי.אשׁור.עד.ה.יום.ה.זה          מ              1     ?     
896   מ.נערי.עד.ה.יום.ה.זה                    מ              1     ?     
903   מ.עולם.עד.עולם                          מ              1     ?     
905   מ.ערב.עד.בקר                            מ              2     ?     
955   מן.ה.בטן.עד.יום.מותו                    מן             1     ?     
964   מן.ה.יום.ה.ראשׁון.עד.ה.יום.ה.אחרון      מן             1     ?     
973   מני.בטן                                 מני            1     ?     
974   מני.רחם                                 מני            1     ?     
1009  עד.יום.אחד.ל.ה.חדשׁ.ה.ראשׁון            עד             1     ?     
1069  ראשׁ.ה.אשׁמרת.ה.תיכונה                  ראשׁ           1     ?     
1095  שׁבעת.ימים.ו.שׁבעת.ימים.ארבעה.עשׂר.יום  ימים.ימים      1     ?     
1105  שׁלשׁ.שׁנים.שׁנה.אחרי.שׁנה              שׁנים          1     ?     
1164  תמיד.מ.רשׁית.ה.שׁנה.ו.עד.אחרית.שׁנה     תמיד           1     ?     

                                                                                                                                                                                                                      note  \
13    One case in Job with 1cs suffix אחרי that looks debatable                                                                                                                                                              
159   shouldn’t הלאה be considered a co-head here?                                                                                                                                                                           
245   how to handle the second prep עד, here? Similar to הלאה. They could be treated as specifications as they probably are in the current dataset. This is not necessarily a bad encoding.                                  
252   this does not appear adverbial but seems to be modifying ימי…; tell-tale sign is that it immediately follows a PP/NP, other questionable cases could be found in this way                                              
651   this TP has quantified ימים with waw separating second part; this is typically represented as two heads, but not here. Why? Is it a problem?                                                                           
802   There is more inconsistency here with the second prep specification. Is it or is it not a “head” element? This is done differently here than it is elsewhere…                                                          
830   I wonder whether this really is temporal in nature? this needs more investigation. This case was marked as a statistically rare head term for TPs in the statistical TP head tests.                                    
864   Is this different from coordinated phrases with two preps?                                                                                                                                                             
876   Only one prep counted due to no waw coordination. Have a look at two examples above, from Jer 36:2, where a near-verbatim phrase is deployed without the waw                                                           
896   missing עד; I marked this case, but there are some other cases that I indeed didn’t mark…                                                                                               

And for nheads...

In [12]:
nhead_anno[nhead_anno['mark']=='?']

token  \
75   ב.דמי.ימי                                                  
84   ב.ה.חגים.ו.ב.ה.חדשׁים.ו.ב.ה.שׁבתות.ב.כל.מועדי.בית.ישׂראל   
111  ב.ה.חיים.מספר.ימי.חיי.הבלו                                 
146  ב.ה.ליל                                                    
220  ב.ה.שׁנים.עד.שׁנת.ה.יבל                                    
273  ב.יום.ו.עד.ה.יום.ה.זה                                      
320  ב.ימי                                                      
358  ב.ימי.קציר.ב.ה.ראשׁנים..תחלת.קציר.שׂערים                   
371  ב.יצקתו                                                    
398  ב.עד.יומם                                                  
399  ב.עוד.ה.יום                                                
400  ב.עוד.כברת.ארץ                                             
402  ב.עוד.שׁלשׁה.חדשׁים.ל.ה.קציר                               
403  ב.עוד.שׁלשׁת.ימים                                          
404  ב.עוד.שׁנתים.ימים                                          
405  ב.עוד.שׁשׁים.ו.חמשׁ.שׁנה                                   
410  ב.עשׂרים.ו.ארבעה.ל.ה.חדשׁ                                  
432  ב.ראשׁי.חדשׁיכם                                            
450  ב.שׁלשׁה.עשׂר.בו.ו.ב.ארבעה.עשׂר.בו                         
558  כ.אתמול.שׁלשׁום                                            
561  כ.ה.יום                                                    
562  כ.ה.יום.ה.זה                                               
570  כ.יום.תמים                                                 
573  כ.מה.שׁנים                                                 
576  כ.מעט.רגע                                                  
585  ל.ארך.ימים                                                 
589  ל.דר.ו.דר.עד.עולם                                          
609  ל.ה.מצער                                                   
623  ל.ימים                                                     
677  ל.פניהם                                                    
688  ל.רוח.ה.יום                                                
704  מ.בטן                                                      
705  מ.בקר                                                      
707  מ.די.חדשׁ.ב.חדשׁו.ו.מ.די.שׁבת.ב.שׁבתו                      
708  מ.די.שׁנה.ב.שׁנה                                           
745  מ.מעי.אמי                                                  
826  עד.אן                                                      
827  עד.אנה                                                     
877  עד.ערב.בקר.אלפים.ו.שׁלשׁ.מאות                              

         (n)heads_token  freq mark  \
75   דמי                 1     ?     
84   חגים.חדשׁים.שׁבתות  1     ?     
111  חיים                1     ?     
146  ליל                 1     ?     
220  שׁנים               1     ?     
273  יום.יום             1     ?     
320  ימי                 4     ?     
358  ימי                 1     ?     
371  יצקתו               1     ?     
398  עד                  1     ?     
399  יום                 1     ?     
400  כברת                1     ?     
402  חדשׁים              1     ?     
403  ימים                3     ?     
404  שׁנתים              2     ?     
405  שׁנה                1     ?     
410  עשׂרים.ארבעה        1     ?     
432  חדשׁיכם             1     ?     
450  שׁלשׁה.ארבעה        1     ?     
558  אתמול               2     ?     
561  יום                 8     ?     
562  יום                 8     ?     
570  יום                 1     ?     
573  מה                  1     ?     
576  רגע                 2     ?     
585  ארך                 2     ?     
589  דר.דר               1     ?     
609  מצער                1     ?     
623  ימים                1     ?     
677  פניהם               1     ?     
688  רוח                 1     ?     
704  בטן                 3     ?     
705  בקר                 1     ?     
707  די.די               1     ?     
708  די                  2     ?     
745  מעי                 1     ?     
826  אן                  1     ?     
827  אנה   

# Conclusions

This analysis manually looked over 1164 head to token phrase mappings and 888 nhead to token mappings. Many of the problematic cases have been fixed by exporting new phrase functions, adjusting the heads pipeline, or planning for new chunk objects that can handle complexity above/below the phrase level. Some more complex cases with heads still need to be adjusted (see bold notes in discussion on fixes above).